File: backend/app/models/marketplace.py
"""
Prime Business Network – Marketplace Models.

Handles member-to-member product and service listings.
"""

from __future__ import annotations

import enum
import uuid
from decimal import Decimal
from typing import List

from sqlalchemy import Boolean, ForeignKey, String, Text, Enum, Numeric, UniqueConstraint
from sqlalchemy.dialects.postgresql import UUID, JSONB
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.models.base import Base, TimestampMixin


class ListingCategory(str, enum.Enum):
    PRODUCT = "product"
    SERVICE = "service"
    CONSULTATION = "consultation"


class ListingStatus(str, enum.Enum):
    ACTIVE = "active"
    PAUSED = "paused"
    SOLD_OUT = "sold_out"
    FLAGGED = "flagged"
    REMOVED = "removed"


class MarketplaceListing(Base, TimestampMixin):
    __tablename__ = "marketplace_listings"

    seller_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True
    )
    title: Mapped[str] = mapped_column(String(255), nullable=False)
    description: Mapped[str] = mapped_column(Text, nullable=False)
    category: Mapped[ListingCategory] = mapped_column(
        Enum(ListingCategory), default=ListingCategory.PRODUCT, nullable=False, index=True
    )
    industry_category_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("industry_categories.id"), nullable=False, index=True
    )
    
    # Pricing
    regular_price: Mapped[Decimal | None] = mapped_column(Numeric(15, 2), nullable=True)
    member_price: Mapped[Decimal | None] = mapped_column(Numeric(15, 2), nullable=True)
    currency: Mapped[str] = mapped_column(String(10), default="LKR")
    price_note: Mapped[str | None] = mapped_column(String(100), nullable=True)
    
    # Media & Display
    image_urls: Mapped[List[str]] = mapped_column(JSONB, default=list, nullable=False)
    is_featured: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False)
    status: Mapped[ListingStatus] = mapped_column(
        Enum(ListingStatus), default=ListingStatus.ACTIVE, nullable=False, index=True
    )
    
    # Contact Info
    whatsapp_number: Mapped[str | None] = mapped_column(String(20), nullable=True)
    contact_email: Mapped[str | None] = mapped_column(String(255), nullable=True)
    contact_phone: Mapped[str | None] = mapped_column(String(20), nullable=True)

    # Metrics
    view_count: Mapped[int] = mapped_column(default=0, nullable=False)
    interest_count: Mapped[int] = mapped_column(default=0, nullable=False)

    # Relationships
    seller = relationship("User", lazy="joined")
    industry = relationship("IndustryCategory", lazy="joined")
    interests = relationship("MarketplaceInterest", back_populates="listing", cascade="all, delete-orphan")

    def __repr__(self) -> str:
        return f"<MarketplaceListing id={self.id} title={self.title}>"


class InterestStatus(str, enum.Enum):
    PENDING = "pending"
    DEAL_CONFIRMED = "deal_confirmed"
    CANCELLED = "cancelled"


class MarketplaceInterest(Base, TimestampMixin):
    __tablename__ = "marketplace_interests"

    listing_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("marketplace_listings.id", ondelete="CASCADE"), nullable=False, index=True
    )
    interested_user_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True
    )
    message: Mapped[str | None] = mapped_column(Text, nullable=True)
    is_read: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False)
    status: Mapped[InterestStatus] = mapped_column(
        Enum(InterestStatus), default=InterestStatus.PENDING, nullable=False, index=True
    )
    business_value: Mapped[Decimal | None] = mapped_column(Numeric(15, 2), default=0, nullable=True)

    # Relationships
    listing = relationship("MarketplaceListing", back_populates="interests")
    interested_user = relationship("User", lazy="joined")

    __table_args__ = (
        UniqueConstraint("listing_id", "interested_user_id", name="unique_member_interest"),
    )

    def __repr__(self) -> str:
        return f"<MarketplaceInterest listing={self.listing_id} user={self.interested_user_id}>"

File: mobile/lib/models/marketplace.dart
import 'package:json_annotation/json_annotation.dart';

part 'marketplace.g.dart';

double? _parseDouble(dynamic value) {
  if (value == null) return null;
  if (value is num) return value.toDouble();
  if (value is String) return double.tryParse(value);
  return null;
}

enum ListingCategory {
  @JsonValue('product')
  product,
  @JsonValue('service')
  service,
  @JsonValue('consultation')
  consultation,
}

enum ListingStatus {
  @JsonValue('active')
  active,
  @JsonValue('paused')
  paused,
  @JsonValue('sold_out')
  sold_out,
  @JsonValue('flagged')
  flagged,
  @JsonValue('removed')
  removed,
}

@JsonSerializable(fieldRename: FieldRename.snake)
class MarketplaceListing {
  final String id;
  final String sellerId;
  final String title;
  final String description;
  final ListingCategory category;
  final String industryCategoryId;
  
  @JsonKey(fromJson: _parseDouble)
  final double? regularPrice;
  @JsonKey(fromJson: _parseDouble)
  final double? memberPrice;
  final String currency;
  final String? priceNote;
  
  final List<String> imageUrls;
  final bool isFeatured;
  final ListingStatus status;
  
  final String? whatsappNumber;
  final String? contactEmail;
  final String? contactPhone;
  
  final int viewCount;
  final int interestCount;
  final bool isApproved;
  final String? rejectionReason;
  final DateTime createdAt;
  final DateTime updatedAt;
  
  final String? sellerName;
  final String? industryName;

  MarketplaceListing({
    required this.id,
    required this.sellerId,
    required this.title,
    required this.description,
    required this.category,
    required this.industryCategoryId,
    this.regularPrice,
    this.memberPrice,
    required this.currency,
    this.priceNote,
    required this.imageUrls,
    required this.isFeatured,
    required this.status,
    this.whatsappNumber,
    this.contactEmail,
    this.contactPhone,
    required this.viewCount,
    required this.interestCount,
    required this.isApproved,
    this.rejectionReason,
    required this.createdAt,
    required this.updatedAt,
    this.sellerName,
    this.industryName,
  });

  factory MarketplaceListing.fromJson(Map<String, dynamic> json) => _$MarketplaceListingFromJson(json);
  Map<String, dynamic> toJson() => _$MarketplaceListingToJson(this);
}

enum InterestStatus {
  @JsonValue('pending')
  pending,
  @JsonValue('deal_confirmed')
  deal_confirmed,
  @JsonValue('cancelled')
  cancelled,
}

@JsonSerializable(fieldRename: FieldRename.snake)
class MarketplaceInterest {
  final String id;
  final String listingId;
  final String interestedUserId;
  final String interestedUserName;
  final String? message;
  final bool isRead;
  final InterestStatus status;
  @JsonKey(fromJson: _parseDouble)
  final double? businessValue;
  final DateTime createdAt;

  MarketplaceInterest({
    required this.id,
    required this.listingId,
    required this.interestedUserId,
    required this.interestedUserName,
    this.message,
    required this.isRead,
    required this.status,
    this.businessValue,
    required this.createdAt,
  });

  factory MarketplaceInterest.fromJson(Map<String, dynamic> json) => _$MarketplaceInterestFromJson(json);
  Map<String, dynamic> toJson() => _$MarketplaceInterestToJson(this);
}
